trinity/sync/common/headers.py

```diff
@@ -0,0 +1,732 @@
+        else:
+            skip = self._skip_length
+
+        if max_headers is None:
+            header_limit = peer.max_headers_fetch
+        else:
+            header_limit = min(max_headers, peer.max_headers_fetch)
+
+        try:
+            self.logger.debug("Requsting chain of headers from %s starting at #%d", peer, start_at)
+
+            headers = await self.wait(peer.requests.get_block_headers(
+                start_at,
+                header_limit,
+                skip,
+                reverse=False,
+            ))
+
+            self.logger.trace('sync received new headers', headers)
+        except OperationCancelled:
+            self.logger.info("Skeleteon sync with %s cancelled", peer)
+            return tuple()
+        except TimeoutError:
+            self.logger.warn("Timeout waiting for header batch from %s, aborting sync", peer)
+            await peer.disconnect(DisconnectReason.timeout)
+            return tuple()
+        except ValidationError as err:
+            self.logger.warn(
+                "Invalid header response sent by peer %s disconnecting: %s",
+                peer, err,
+            )
+            await peer.disconnect(DisconnectReason.subprotocol_error)
+            return tuple()
+
+        if not headers:
+            self.logger.info("Got no new headers from %s, exiting skeleton sync", peer)
+            return tuple()
+        else:
+            return headers
+
+    async def _get_missing_tail(
+            self,
+            headers: Tuple[BlockHeader, ...]) -> AsyncIterator[BlockHeader]:
+        """
+        We only want headers that are missing, so we iterate over the list
+        until we find the first missing header, after which we return all of
+        the remaining headers.
+        """
+        iter_headers = iter(headers)
+        for header in iter_headers:
+            is_present = await self.wait(self._db.coro_header_exists(header.hash))
+            if is_present:
+                self.logger.debug("Discarding header that we already have: %s", header)
+            else:
+                yield header
+                break
+
+        for header in iter_headers:
+            yield header
+
+    async def _log_ancester_failure(self, peer: TChainPeer, first_header: BlockHeader) -> None:
+        self.logger.info("Unable to find common ancestor betwen our chain and %s", peer)
+        parent_num = first_header.block_number - 1
+        try:
+            parent = await self._db.coro_get_canonical_block_header_by_number(parent_num)
+        except HeaderNotFound as exc:
+            self.logger.debug("Could not find any header at #%d: %s", parent_num, exc)
+            parent = None
+        self.logger.debug(
+            "%s returned starting header %s, whose parent %s is not in our DB. "
+            "Instead at #%d, is header %s",
+            peer,
+            first_header,
+            first_header.parent_hash,
+            parent_num,
+            parent,
```

Comment:
I'll update the naming to make it a little clearer what's happening. `parent` may or may not be none, which is useful to see in the log, and hopefully readable-ish eg~ "at #123456 is header None". I will change the name to `previous` since it's explicitly *not* the parent of the first header that we're printing, it's just whichever block is in the block number previous.